File: ieee2022/java/src/DisjointSubsets/UpTreeSize.java
package DisjointSubsets;

import java.util.ArrayList;
import java.util.Random;

public class UpTreeSize<T> {

    public T index;
    public UpTreeSize<T> parent;
    public int size;

    public UpTreeSize(T index) {
        this.index = index;
        this.parent = this;
        this.size = 1;
    }

    public UpTreeSize<T> find() {
        UpTreeSize<T> root = this;
        while (root.parent != root) {
            root = root.parent;
        }

        UpTreeSize<T> active = this;
        UpTreeSize<T> next;
        while (active != root) {
            next = active.parent;
            active.parent = root;
            active = next;
        }

        return root;
    }

    public UpTreeSize<T> findPathSplit() {
        UpTreeSize<T> active = this;
        UpTreeSize<T> temp;
        while (active.parent != active) {
            temp = active.parent;
            active.parent = temp.parent;
            active = temp;
        }
        return active;
    }

    public UpTreeSize<T> findPathHalving() {
        UpTreeSize<T> active = this;
        UpTreeSize<T> temp;
        while (active.parent != active) {
            active.parent = active.parent.parent;
            active = active.parent;
        }
        return active;
    }

    public UpTreeSize<T> getRoot() {
        return this.find();
    }

    public static <T> void union(UpTreeSize<T> a, UpTreeSize<T> b) {
        UpTreeSize<T> aRoot = a.getRoot();
        UpTreeSize<T> bRoot = b.getRoot();
        if (aRoot.size < bRoot.size) {
            aRoot.parent = bRoot;
            bRoot.size += aRoot.size;
        } else if(bRoot.size < aRoot.size) {
            bRoot.parent = aRoot;
            aRoot.size += bRoot.size;
        } else {
            aRoot.parent = bRoot;
            bRoot.size += aRoot.size;
        }
    }

    public static <T> UpTreeSize<T> unionWithReturn(UpTreeSize<T> a, UpTreeSize<T> b) {
        UpTreeSize<T> aRoot = a.getRoot();
        UpTreeSize<T> bRoot = b.getRoot();
        if (aRoot.size < bRoot.size) {
            aRoot.parent = bRoot;
            bRoot.size += aRoot.size;
            return bRoot;
        }
        if(bRoot.size < aRoot.size) {
            bRoot.parent = aRoot;
            aRoot.size += bRoot.size;
            return aRoot;
        }
        aRoot.parent = bRoot;
        bRoot.size += aRoot.size;
        return bRoot;
    }

    public static <T> boolean inSameSet(UpTreeSize<T> a, UpTreeSize<T> b) {
        return a.getRoot() == b.getRoot();
    }

    public static void main(String[] args) {
        Random random = new Random();
        ArrayList<UpTreeSize<Integer>> trees = new ArrayList<>();
        long totalTime = - System.currentTimeMillis();
        for (int i = 0; i < 1000000; i++) {
            UpTreeSize<Integer> tree = new UpTreeSize<>(i);
            trees.add(tree);
            union(tree, trees.get(random.nextInt(trees.size())));
        }
        System.out.println((totalTime + System.currentTimeMillis()) / 1000.0);
    }
}

File: ieee2022/python/disjoint_subsets/up_tree_size.py
import random
import time


class UpTreeSize:

    def __init__(self, index):
        self.index = index
        self.parent = self
        self.size = 1

    def find(self):
        root = self
        while root.parent != root:
            root = root.parent

        active = self
        while active != root:
            next = active.parent
            active.parent = root
            active = next

        return root

    def find_path_split(self):
        active = self
        while active.parent != active:
            active, active.parent = active.parent, active.parent.parent
        return active

    def find_path_halving(self):
        active = self
        while active.parent != active:
            active.parent = active.parent.parent
            active = active.parent
        return active

    get_root = find

    @staticmethod
    def union(a, b):
        a_root = a.get_root()
        b_root = b.get_root()
        if a_root.size < b_root.size:
            a_root.parent = b_root
            b_root.size += a_root.size
        elif b_root.size < a_root.size:
            b_root.parent = a_root
            a_root.size += b_root.size
        else:
            a_root.parent = b_root
            b_root.size += a_root.size

    @staticmethod
    def union_with_return(a, b):
        a_root = a.get_root()
        b_root = b.get_root()
        if a_root.size < b_root.size:
            a_root.parent = b_root
            b_root.size += a_root.size
            return b_root
        if b_root.size < a_root.size:
            b_root.parent = a_root
            a_root.size += b_root.size
            return a_root
        a_root.parent = b_root
        b_root.size += a_root.size
        return b_root

    @staticmethod
    def in_same_set(a, b):
        return a.get_root() == b.get_root()


if __name__ == '__main__':
    trees = []
    total_time = 0
    total_time -= time.time()
    for i in range(1000000):
        tree = UpTreeSize(i)
        trees.append(tree)
        UpTreeSize.union(tree, random.choice(trees))
    total_time += time.time()
    print(total_time)
